File: ThunderBasics/UIView+Pop.m
//
//  UIView+Pop.m
//  ThunderStorm
//
//  Created by Andrew Hart on 20/11/2013.
//  Copyright (c) 2013 3 SIDED CUBE. All rights reserved.
//

#import "UIView+Pop.h"

@implementation UIView (Pop)

- (void)popIn
{
    CAKeyframeAnimation *animation = [CAKeyframeAnimation
                                      animationWithKeyPath:@"transform"];
    
    CATransform3D scale1 = CATransform3DMakeScale(0.1, 0.1, 1);
    CATransform3D scale2 = CATransform3DMakeScale(1.1, 1.1, 1);
    CATransform3D scale3 = CATransform3DMakeScale(0.95, 0.95, 1);
    CATransform3D scale4 = CATransform3DMakeScale(1.0, 1.0, 1);
    
    NSArray *frameValues = @[[NSValue valueWithCATransform3D:scale1],
                             [NSValue valueWithCATransform3D:scale2],
                             [NSValue valueWithCATransform3D:scale3],
                             [NSValue valueWithCATransform3D:scale4]];
    [animation setValues:frameValues];
    
    animation.fillMode = kCAFillModeForwards;
    animation.removedOnCompletion = NO;
    animation.duration = 0.9;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    
    [self.layer addAnimation:animation forKey:nil];
}

- (void)popOut
{
    CAKeyframeAnimation *animation = [CAKeyframeAnimation
                                      animationWithKeyPath:@"transform"];
    
    CATransform3D scale1 = CATransform3DMakeScale(1.0, 1.0, 1);
    CATransform3D scale3 = CATransform3DMakeScale(0.01, 0.01, 1);
    
    NSArray *frameValues = @[[NSValue valueWithCATransform3D:scale1],
                             [NSValue valueWithCATransform3D:scale3]];
    [animation setValues:frameValues];
    
    animation.fillMode = kCAFillModeForwards;
    animation.removedOnCompletion = NO;
    animation.duration = 0.2;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    
    [self.layer addAnimation:animation forKey:nil];
}

@end

File: Sources/ThunderBasics/UIView+Pop.swift
//
//  UIView+Pop.swift
//  ThunderBasics
//
//  Created by Simon Mitchell on 05/11/2018.
//  Copyright © 2018 threesidedcube. All rights reserved.
//

import UIKit

extension UIView {
    
    /// Pops a view inwards, the view will contract inwards then expand back to its original size
    public func popIn() {
        
        let animation = CAKeyframeAnimation(keyPath: "transform")
        let transform1 = CATransform3DMakeScale(0.1, 0.1, 1)
        let transform2 = CATransform3DMakeScale(1.1, 1.1, 1)
        let transform3 = CATransform3DMakeScale(0.95, 0.95, 1)
        let transform4 = CATransform3DMakeScale(1.0, 1.0, 1)
        
        animation.values = [
            transform1, transform2, transform3, transform4
        ]
        
        animation.fillMode = .forwards
        animation.isRemovedOnCompletion = false
        animation.duration = 0.9
        animation.timingFunction = CAMediaTimingFunction(name: .easeOut)
        
        layer.add(animation, forKey: nil)
    }
    
    /// Pops a view outwards, the view will expand outwards then contract back to it's original size
    public func popOut() {
        
        let animation = CAKeyframeAnimation(keyPath: "transform")
        let transform1 = CATransform3DMakeScale(1.0, 1.0, 1)
        let transform2 = CATransform3DMakeScale(0.01, 0.01, 1)
        animation.values = [
            transform1, transform2
        ]
        
        animation.fillMode = .forwards
        animation.isRemovedOnCompletion = false
        animation.duration = 0.2
        animation.timingFunction = CAMediaTimingFunction(name: .easeOut)
        
        layer.add(animation, forKey: nil)
    }
}
